### PR TITLE
executor,table: fix 'show create table' for the temporary table

### LINKED ISSUE #24788: SHOW CREATE TABLE nil pointer dereference for temporary table
## Development Task

```
mysql> create global temporary table t3 (i int primary key, j int) on commit delete rows;
Query OK, 0 rows affected (0.01 sec)

mysql> show create table t3;
ERROR 1105 (HY000): runtime error: invalid memory address or nil pointer dereference
```

On the master branch

```
[err="runtime error: invalid memory address or nil pointer dereference                                                                                                                                             
github.com/pingcap/tidb/executor.(*recordSet).Next.func1                                                                                                                                                           
        /home/genius/project/src/github.com/pingcap/tidb/executor/adapter.go:138                                                                                                                                   
runtime.gopanic                                                                                                                                                                                                    
        /snap/go/7360/src/runtime/panic.go:965                                                                                                                                                                     
runtime.panicmem                                                                                                                                                                                                   
        /snap/go/7360/src/runtime/panic.go:212                                                                                                                                                                     
runtime.sigpanic                                                                                                                                                                                                   
        /snap/go/7360/src/runtime/signal_unix.go:734                                                                                                                                                               
github.com/pingcap/tidb/meta/autoid.Allocators.Get                                                                                                                                                                 
        /home/genius/project/src/github.com/pingcap/tidb/meta/autoid/autoid.go:161                                                                                                                                 
github.com/pingcap/tidb/executor.ConstructResultOfShowCreateTable                                                                                                                                                  
        /home/genius/project/src/github.com/pingcap/tidb/executor/show.go:972                                                                                                                                      
github.com/pingcap/tidb/executor.(*ShowExec).fetchShowCreateTable                                                                                                                                                  
        /home/genius/project/src/github.com/pingcap/tidb/executor/show.go:1094                                                                                                                                     
github.com/pingcap/tidb/executor.(*ShowExec).fetchAll                                                                                                                                                              
        /home/genius/project/src/github.com/pingcap/tidb/executor/show.go:138                                                                                                                                      
github.com/pingcap/tidb/executor.(*ShowExec).Next                                                                                                                                                                  
        /home/genius/project/src/github.com/pingcap/tidb/executor/show.go:101                                                                                                                                      
github.com/pingcap/tidb/executor.Next                                                                                                                                                                              
        /home/genius/project/src/github.com/pingcap/tidb/executor/executor.go:283                                                                                                                                  
github.com/pingcap/tidb/executor.(*recordSet).Next                                                                                                                                                                 
        /home/genius/project/src/github.com/pingcap/tidb/executor/adapter.go:142                                                                                                                                   
github.com/pingcap/tidb/server.(*tidbResultSet).Next                                                                                                                                                               
        /home/genius/project/src/github.com/pingcap/tidb/server/driver_tidb.go:304                                                                                                                                 
github.com/pingcap/tidb/server.(*clientConn).writeChunks                                                                                                                                                           
        /home/genius/project/src/github.com/pingcap/tidb/server/conn.go:1881                                                                                                                                       
github.com/pingcap/tidb/server.(*clientConn).writeResultset                                                                                                                                                        
        /home/genius/project/src/github.com/pingcap/tidb/server/conn.go:1829                                                                                                                                       
github.com/pingcap/tidb/server.(*clientConn).handleStmt                                                                                                                                                            
        /home/genius/project/src/github.com/pingcap/tidb/server/conn.go:1723                                                                                                                                       
github.com/pingcap/tidb/server.(*clientConn).handleQuery                                                                                                                                                           
        /home/genius/project/src/github.com/pingcap/tidb/server/conn.go:1569                                                                                                                                       
github.com/pingcap/tidb/server.(*clientConn).dispatch                                                                                                                                                              
        /home/genius/project/src/github.com/pingcap/tidb/server/conn.go:1103                                                                                                                                       
github.com/pingcap/tidb/server.(*clientConn).Run                                                                                                                                                                   
        /home/genius/project/src/github.com/pingcap/tidb/server/conn.go:861                                                                                                                                        
github.com/pingcap/tidb/server.(*Server).onConn                                                                                                                                                                    
        /home/genius/project/src/github.com/pingcap/tidb/server/server.go:482                                                                                                                                      
runtime.goexit                                                                                                                                                                                                     
        /snap/go/7360/src/runtime/asm_amd64.s:1371" 
```

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve?

Issue Number: close #24788 #24752

Problem Summary:

### What is changed and how it works?

What's Changed:

Support `show create table` statement for the temporary table.

How it Works:

- Add `GLOBAL TEMPORARY ... ON COMMIT DELETE ROWS` to the result
- Avoid nil allocator for the temporary table in the `show create table` statement.

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### CODE REVIEW

File: table/tables/tables.go [hunk 1]
```diff
@@ -1374,8 +1374,11 @@ func (t *TableCommon) Allocators(ctx sessionctx.Context) autoid.Allocators {
 	} else if ctx.GetSessionVars().IDAllocator == nil {
 		// Use an independent allocator for global temporary tables.
 		if t.meta.TempTableType == model.TempTableGlobal {
-			alloc := ctx.GetSessionVars().GetTemporaryTable(t.meta).GetAutoIDAllocator()
-			return autoid.Allocators{alloc}
+			if alloc := ctx.GetSessionVars().GetTemporaryTable(t.meta).GetAutoIDAllocator(); alloc != nil {
+				return autoid.Allocators{alloc}
+			}
+			// If the session is not in a txn, for example, in "show create table", use the original allocator.
+			// Otherwise the would be a nil pointer dereference,
```
Comment (wjhuang2016) [inline table/tables/tables.go:1381]: s/,/.

Comment (tiancaiamao) [reply table/tables/tables.go]: Done

**Comment by tiancaiamao** [comment]
PTAL @djshow832 @wjhuang2016

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -2225,7 +2225,7 @@ func handleTableOptions(options []*ast.TableOption, tbInfo *model.TableInfo) err
 			// We don't handle charset and collate here since they're handled in `getCharsetAndCollateInTableOption`.
 		case ast.TableOptionEngine:
 			if tbInfo.TempTableType != model.TempTableNone {
-				if op.StrValue != "" && !strings.EqualFold(op.StrValue, "memory") {
+				if op.StrValue != "" && !strings.EqualFold(op.StrValue, "memory") && !strings.EqualFold(op.StrValue, "innodb") {
```
Comment (djshow832) [inline ddl/ddl_api.go:2228]: We don't support innodb.

**Pull request merged**